app/commands/user_track/generate_concept_exercise_mapping.rb

```diff
@@ -0,0 +1,35 @@
+class UserTrack
+  class GenerateConceptExerciseMapping
+    include Mandate
+
+    initialize_with :user_track
+
+    def call
+      concept_exercise_mapping =
+        user_track.
+          track.
+          exercises.
+          # includes(:taught_concepts, :prerequisites). <-- How to optimize? raises error d/t practice exercise
+          each_with_object({}) do |exercise, hash|
+            concepts = exercise.practice_exercise? ? exercise.prerequisites : exercise.taught_concepts
+
+            concepts.to_a.each do |concept|
+              slug = concept.slug
+              hash[slug] ||= { exercises: 0, exercises_completed: 0 }
+              hash[slug][:exercises] += 1
+            end
+          end
+
+      concept_exercise_mapping.tap do |hash|
+        user_track.solutions.completed.each do |solution|
+          exercise = solution.exercise
+          concepts = exercise.practice_exercise? ? exercise.prerequisites : exercise.taught_concepts
+          concepts.to_a.each do |concept|
+            slug = concept.slug
+            hash[slug][:exercises_completed] += 1
+          end
+        end
+      end
+    end
```

Comment:
So you can probably do all this in four SQL queries using grouping (you could in one, but we'd probably need to break into actual SQL for that, and it's nicer to stick with AR until we want to optimise)

```ruby
prereq_counts = ExercisePrerequisite.where(exercise_id: Track.practice_exercises).group(:concept_id).count
taught_counts = ExerciseTaughtConcept.where(exercise_id: Track.concept_exercises).group(:concept_id).count

solved_prereq_counts = ExercisePrerequisite.where(exercise: Solution.where(exercise_id: Track.practice_exercises).completed.select(:exercise_id)).group(:concept_id).count
solved_taught_counts = ExerciseTaughtConcept.where(exercise: Solution.where(exercise_id: Track.concept_exercises).completed.select(:exercise_id)).group(:concept_id).count

concept_exercise_mapping = Concept.all.map do |concept|
  {
     exercises: pr_counts[concept_id] + taught_counts[concept_id],
     exercises_completed: solved_prereq_counts[concept_id] + solved_taught_counts[concept_id]
  }
end
```

I've just written this in GitHub so no idea if it works etc :)